### DISCUSSION TITLE
Possible to "unwrap" a Promise result type?

### DISCUSSION THREAD

**Discussion by rchl** [python/typing#1342] [Q&A] [answered]
I have a custom Promise implementation in Python and I'm struggling with an issue where in certain situations the type of the value passed to `.then` is not "unwrapped" from the `Promise` type.

I'll post all the code at the end but here are some examples (keep in mind that it should work more or less like Javascript implementation):

1. In this example the final type of `result` is an `int` as expected

```py
Promise.resolve(None) \
  .then(lambda _: 123)
  .then(lambda result: reveal_type(result))
```

2. In this example the type of final `result` is `Promise[str | int]` and that's an issue because the value passed to `then` should always be an "unwrapped" value. So the whole chain of promises should have been resolved so that the `result` type should no longer be wrapped in `Promise`. The implementation works like that but I'm not sure how to express that with types.

```py
Promise.resolve(None) \
  .then(lambda result: Promise.resolve('abc' or 123)) \
  .then(lambda result: reveal_type(result))
```

The problem seems to be specific to `Promise.resolve` which uses different `TypeVar` than the main `Promise` class. Reasoning for that is in this issue: https://github.com/microsoft/pyright/issues/3963

So my question is: Is it possible to achieve the expected types behavior in Python?

The simplified Promise implementation is as follows:

<details>
<summary>promise.py</summary>

```py
import functools
import threading
from typing import Callable, Generic, List, Protocol, TypeVar, Union

T = TypeVar('T')
S = TypeVar('S')
T_contra = TypeVar('T_contra', contravariant=True)
TResult = TypeVar('TResult')


class ResolveFunc(Protocol[T_contra]):
    def __call__(self, resolve_value: T_contra) -> None:
        ...


FullfillFunc = Callable[[T], Union[TResult, 'Promise[TResult]']]
ExecutorFunc = Callable[[ResolveFunc[T]], None]


class Promise(Generic[T]):

    @staticmethod
    def resolve(resolve_value: S) -> 'Promise[S]':

        def executor_func(resolve_fn: ResolveFunc[S]) -> None:
            resolve_fn(resolve_value)

        return Promise(executor_func)

    def __init__(self, executor_func: ExecutorFunc[T]) -> None:
        self.resolved = False
        self.mutex = threading.Lock()
        self.callbacks = []  # type: List[ResolveFunc[T]]
        executor_func(lambda resolve_value=None: self._do_resolve(resolve_value))

    def then(self, onfullfilled: FullfillFunc[T, TResult]) -> 'Promise[TResult]':

        def callback_wrapper(resolve_fn: ResolveFunc[TResult], resolve_value: T) -> None:
            result = onfullfilled(resolve_value)
            # If returned value is a promise then this promise needs to be
            # resolved with the value of returned promise.
            if isinstance(result, Promise):
                result.then(lambda value: resolve_fn(value))
            else:
                resolve_fn(result)

        def sync_wrapper(resolve_fn: ResolveFunc[TResult]) -> None:
            callback_wrapper(resolve_fn, self._get_value())

        def async_wrapper(resolve_fn: ResolveFunc[TResult]) -> None:
            self._add_callback(functools.partial(callback_wrapper, resolve_fn))

        if self._is_resolved():
            return Promise(sync_wrapper)
        return Promise(async_wrapper)

    def _do_resolve(self, new_value: T) -> None:
        # No need to block as we can't change from resolved to unresolved.
        if self.resolved:
            raise RuntimeError("cannot set the value of an already resolved promise")
        with self.mutex:
            self.resolved = True
            self.value = new_value
            for callback in self.callbacks:
                callback(new_value)

    def _add_callback(self, callback: ResolveFunc[T]) -> None:
        with self.mutex:
            self.callbacks.append(callback)

    def _is_resolved(self) -> bool:
        with self.mutex:
            return self.resolved

    def _get_value(self) -> T:
        with self.mutex:
            return self.value
```

</details>

I've also created a [pyright playground link](https://pyright-playground.decorator-factory.su/?gzip=H4sIAF023GMC_61W34vbOBB-918xZB9sg2to-xbY4-DoHoVyHN3cvaTGKLa8EVWkIMubXegf35EtyZKTzXWP9cMSy_Pjm_m-GS07HKXS0A2i0VLyPmHTgd4rSlomHpJOyQPo5yP-BvvxD8I52XFawJ9UUMWaAr6wXhfwt5JaNpIXsHk-0n-JKuAfwaRIkg3curMs3aR5ch8e3OPBpm6k0IpEhvYsLWD68UgUI0LfbtRA0eUr7QeuI4_pCOMlScNJ3wMeSP5I77DCzOHbusBVvk4An5Z2UNcN1lXXWU95V4Ca_OpHwge6BueRw7vf4C8p6ORonrIsMdvdwHnHODeJEJHr0Xa7qWwXthZcASkCObCeupMqrark0xNtBi3VWYCgAgyG0Uz6yhdoY2WWCzTBokZwv_eaaNYcqN7L1tdpC8sWBd6PlXlk91Vqozg_avHVRizeuxPrsMXod6FD5pkd4sz5nEVRPSjhC4oSWrOJKCaY9kRFZmsI22h6cY7GuJUWQ4udviO8p_HXw6DpE37yY1B-kc33LI-tjF52pPneo-W2Argxc4KtNMOwZM17xm3k5LBrSSy22xHulKJuZX2RsDxoiN5TYZshRWd1SNs1hJrcbnAqrdxiqr0IF4S78uqTIscjVS9y7gKcD83LWpjmNoS7lEXocAOfOysPpGw0ANYDgeNUw9gC_INn7kRQ2vagJezoIpSn_sT03njagDJIYaOUkSvrMCkTOFWiGekYh9m2MY9LnMssR3os0bYxwTA4NkNHioK8GC6YIIycx3z1z6L5Za5eYOYK54VV5APVE0VZvgBA3gLBlIS0be2wZP52Ko9EaUZ4toRZBM0JUSFjUzzWuzFqs3wpxmjrhDXkL60mElvNqykY12kgBT1dm4YbfB-lOiqV45LBLsKJIhEi1dDsiXigMN7BXrZoOQj3Vp7V6r4sqiRmKL4OQrMD_aSUVNkKcwip0UvHU0AEEG4W3_Oc1A7Eam7JOD3ztrzAYrBjzYV9bjBlvJ2bFJl0Unk9AhOLnXs-H14vPlzITKSoiRv3utDnJWX-d7EeWGlEIVqv0RBEKEPjNWbaobR_MZOVYdTdIPw8mj745n9EnqhIkht4d_1JEjsQDkz2PodvGDfaedOyMkvvkRJemzvSLbDcTMBqelnBHqVvviLZ2owBfTrSRpsKX5MmJbsmfRMYvVbXYRiNXMm0NB-RAar6_YeP-ZsgdFe4QfrDtK2C055xCp_T1uKGlf-4enUFEeLXAx4v5JNU-B9S1Mef9wHWjXYMAAA%3D) with the implementation and some examples (including the failing one) at the bottom.

**Comment by erictraut** [comment, marked as answer]
Your implementation includes the type subexpression `Union[TResult, Promise[TResult]]`. Any time you create a union between a "naked" TypeVar and a type that uses that that same TypeVar as a type argument, you create the potential for multiple solutions within the constraint solver. Consider the following simplified example:

```python
from typing import TypeVar

T = TypeVar("T")

def func(x: T | list[T]) -> T:
    ...

reveal_type(func([1])) # int or list[int]?
```

In this case, both `int` and `list[int]` are acceptable solutions. However, the solution `int` is probably the one you expect. How does the constraint solver choose between multiple valid answers? There is no standardized technique. Pyright uses a heuristic whereby it computes all of the solutions and "scores" each one. The score attempts to take into account several factors including the number of TypeVars that are solved (or left unsolved) and the "complexity" of the resulting solutions. The "complexity" of a type is difficult to compute because there's subjectivity involved. For example, which is more complex: `str | int` or `Promise[int]`?

In this particular case, the complexity calculation in pyright is producing a result that is somewhat unintuitive. I've tweaked the complexity calculation algorithm slightly to nudge it in the right direction. This will be included in the next release of pyright.

**Reply by rchl** [reply]
Thank you for detailed explanation (as always) and looking forward to trying the new version.
